### DISCUSSION TITLE
Feat request: open Listbox when typing

### DISCUSSION THREAD

**Discussion by robbevp** [tailwindlabs/headlessui#670] [Ideas]
Hi there,

While using the Listbox, I notice that I often start typing when I have the ListboxButton focused ut not yet opened.
At the moment I first need to press enter and then type to find the option I want.

Is it possible to open the Listbox when the user presses any key? (Except a few special cases like esc, tab ...)

**Comment by drusellers** [comment]
I would love to see something like this.

**Comment by RobinMalfait** [comment]
Hey! Thank you for your suggestion!
Much appreciated! 🙏

I like this suggestion, but it is a little bit tricky to get right, let me explain:

We currently have 2 "rendering" strategies, you can use the `unmount={true}` (default) on the `Listbox.Options` so that the `Listbox.Options` get unmounted when the `Listbox` is not open anymore. The other approach is to use `unmount={false}`, which doesn't unmount, but hides the `Listbox.Options`. Why is this important?

If you look at the native `<select>`, then you can start "typing" while the select is focused and still closed. This will also "select" the best match already (it will call your change handlers). You also have to notice that the native `select` will never open in this case. We can implement this when the `unmount={false}` is used, because all the items are rendered in the DOM. However, we cannot implement this (in a straightforward way) when the `unmount={true}` strategy is used, in this case we have to open the `Listbox` first and then start searching (your suggestion).

However this results in 2 different behaviours. We could normalize it to always open, so that it is consistent. But this results in different behaviour between the `Listbox` and the native `select` element, which is a little unfortunate.

In short, while it is possible and a good idea, we have to play with the usability and define what our expectations are.
I am not going to implement it immediately, but I will mark it as a feature request and come back to it later.

Thanks!

**Reply by benface** [reply]
> We could normalize it to always open, so that it is consistent. But this results in different behaviour between the `Listbox` and the native `select` element, which is a little unfortunate.

Personally I think this solution would make a lot of sense and could even be an improvement over the native `select` element’s behavior. The only thing that could be annoying is if the user expects to be able to tab to the next field (assuming they’re filling out a form) right after typing the value they want to select this way; it won’t work because the focus will be trapped in the listbox. But still, it’s easy enough to notice and just press <kbd>Enter</kbd> or <kbd>Space</kbd> or even <kbd>Esc</kbd> (as the value matching what they typed will have been selected, not just focused, right?) and then tab.

**Comment by m-tmvc-krsch** [comment]
Hi, i would try to solve it. It is my first issue in an open source project. Can you tell me what I need to pay special attention to and how to start?
